data-prepper-plugins/s3-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/S3SinkService.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.sink;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.model.record.Record;
+import org.opensearch.dataprepper.model.types.ByteCount;
+import org.opensearch.dataprepper.plugins.sink.accumulator.Buffer;
+import org.opensearch.dataprepper.plugins.sink.accumulator.BufferFactory;
+import org.opensearch.dataprepper.plugins.sink.accumulator.ObjectKey;
+import org.opensearch.dataprepper.plugins.sink.codec.Codec;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.core.client.config.ClientOverrideConfiguration;
+import software.amazon.awssdk.core.retry.RetryPolicy;
+import software.amazon.awssdk.services.s3.S3Client;
+
+/**
+ * Class responsible for create {@link S3Client} object, check thresholds,
+ * get new buffer and write records into buffer.
+ */
+public class S3SinkService {
+
+    private static final Logger LOG = LoggerFactory.getLogger(S3SinkService.class);
+    private final S3SinkConfig s3SinkConfig;
+    private final Lock reentrantLock;
+    private final BufferFactory bufferFactory;
+    private final Codec codec;
+    private final S3Client s3Client;
+    private Buffer currentBuffer;
+    private final int numEvents;
+    private final ByteCount byteCapacity;
+    private final long duration;
+
+    /**
+     * @param s3SinkConfig s3 sink related configuration.
+     * @param codec        parser
+     */
+    public S3SinkService(final S3SinkConfig s3SinkConfig, final BufferFactory bufferFactory, final Codec codec) {
+        this.s3SinkConfig = s3SinkConfig;
+        this.bufferFactory = bufferFactory;
+        this.codec = codec;
+        this.s3Client = createS3Client();
+        reentrantLock = new ReentrantLock();
+
+        numEvents = s3SinkConfig.getThresholdOptions().getEventCount();
+        byteCapacity = s3SinkConfig.getThresholdOptions().getMaximumSize();
+        duration = s3SinkConfig.getThresholdOptions().getEventCollectTimeOut().getSeconds();
```

Comment:
Please update these variable names for better readability, maybe `maxEvents`, `maxBytes`, `maxCollectionDuration`.

Reply:
Updated.